### ISSUE TITLE
Check if GBDT Support Entropy and Gini for Impurity Type

### ISSUE THREAD

**Issue by zhangpengshan** [ShifuML/shifu#369]
Seems exception when run GBDT with entropy impurity type.

**Comment by zhangpengshan** [comment]
No need support:

I checked again why regression tree doesn’t support gini/entropy impurity:
For gini/entropy tree growth, stats are on each label, which means in each label each feature, like label 0 in category A, count is 10, sum is 9.
While for GBDT, label are changed to gradient after the first tree. For example, record A with label 1, but later it is predicted to 0.8, the next label is changed to gradient which is (1-0.8=0.2), 0.2 is not a real label and cannot be stats again.

Below are scikit-learn validation for regression impurity: (MSE is variance in Shifu.)
https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/tree/tree.py#L60
CRITERIA_CLF = {"gini": _criterion.Gini, "entropy": _criterion.Entropy}
CRITERIA_REG = {"mse": _criterion.MSE, "friedman_mse": _criterion.FriedmanMSE,
                "mae": _criterion.MAE}

And Spark MLLib validation for regression impurity:
https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/mllib/tree/configuration/Strategy.scala#L146
      case Classification =>
        require(numClasses >= 2,
          s"DecisionTree Strategy for Classification must have numClasses >= 2," +
          s" but numClasses = $numClasses.")
        require(Set(Gini, Entropy).contains(impurity),
          s"DecisionTree Strategy given invalid impurity for Classification: $impurity." +
          s"  Valid settings: Gini, Entropy")
      case Regression =>
        require(impurity == Variance,
          s"DecisionTree Strategy given invalid impurity for Regression: $impurity." +
          s"  Valid settings: Variance")
